### ISSUE TITLE
How to design for nested props for render function?

### ISSUE THREAD

**Issue by cmeeren** [cmeeren/Feliz.MaterialUI#26]
I am currently updating Feliz.MaterialUI for the new Autocomplete lab component. Is has a couple of render props that accept a `props` object that contain pre-defined props that should be spread to the root component.

For example (JS):

```jsx
renderInput={props => (
    <TextField
      {...props},
      label="Foo"
    />
  )}
```

Based on the above, I figured that I could just model the input `props` as `IReactProperty []` so that I can simply `yield!` them:

```f#
autocomplete.renderInput (fun props ->
  Mui.textField [
    yield! props
    textField.label "Foo"
  ]
)
```

That works fine, but then I came across the following JS which uses TextField's `inputProps`. This prop is part of the `props` object, so `props.inputProps` must then be spread explicitly, before passing other props to `inputProps`:

```jsx
renderInput={props => (
    <TextField
      {...props},
      label="Foo"
      inputProps={{
        ...props.inputProps,
        // ... other input props
      }}
    />
  )}
```

Without the `...props.inputProps,` line, `inputProps` would only contain the custom props, not the props from the input `props`.

I'm a bit lost as to how to design for this. Any suggestions are appreciated.

One option is to model the input `props` as an object:

```f#
type RenderInputProps =
  abstract rootProps: IReactProperty []
  abstract inputProps: RenderInputInputProps

and RenderInputInputProps =
  abstract rootProps: IReactProperty []
  // And further levels if necessary, though I don't think so
```

The F# usage would then be:

```f#
autocomplete.renderInput (fun props ->
  Mui.textField [
    yield! props.rootProps
    textField.label "Foo"
    textField.inputProps [
      yield! props.inputProps.rootProps
      // other input props
    ]
  ]
)
```

**Comment by cmeeren** [comment]
@Zaid-Ajaj, you have previously helped the API design greatly - do you have any ideas here?

**Comment by Zaid-Ajaj** [comment]
> Based on the above, I figured that I could just model the input props as IReactProperty [] so that I can simply yield! 

> That works fine

Does it though or do you mean it only compiles fine? 

I don't think that `props` value is `IReactProperty list` because it is already a transformed `props` *object literal* and not a F# list anymore which means the static type is incorrect with the run-time type (I guess list operations will fail when applied to the object).

This is because the syntax the following:
```js
<TextField
      {...props},
      label="Foo"
      inputProps={{
        ...props.inputProps,
        anotherProp={true}
      }}
    />
```
Compiles down to:
```js
React.createElement(
  TextField, 
  Object.assign({ }, props, { 
    label: "Foo", 
    inputProps: Object.assign({ }, props.inputProps, { anotherProp: true }
  })
)
```
One solution I can think of is re-create the list from the object:
```fs
type autocomplete = 
  static member renderText (renderer: IReactProperty list -> IReactProperty list -> ReactElement) = 
    let wrappedRenderer props = 
      let rootProps = createPropertyList props 
      let inputProps = createPropertyList (props?inputProps)
      renderer rootProps inputProps
      
    unbox<IReactProperty> ("renderText", wrappedRenderer)
```
where `createPropertyList` is a function that takes an object literal and makes a F# list of it by reading the keys and mapping them into tuples of `key * props[key]`

Then this also solves your nesting problem because both `props` and `props.inputProps` are now two parameters of `autocomplete.renderText`:
```fs
autocomplete.renderInput (fun props inputProps ->
  Mui.textField [
    yield! props 
    textField.label "Foo"
    textField.inputProps [
      yield! inputProps
      // other input props
    ]
  ])
```
I hope it is clear, let me know otherwise!

**Comment by cmeeren** [comment]
> I don't think that `props` value is `IReactProperty list` because it is already a transformed `props` _object literal_ and not a F# list anymore which means the static type is incorrect with the run-time type (I guess list operations will fail when applied to the object).

Yes, and I am re-creating the list using `Object.entries` similar to what you suggested in https://github.com/cmeeren/Feliz.MaterialUI/issues/2#issuecomment-527092374. I was thinking more about the design here, but I could have mentioned that I have that particular part under control.

> One solution I can think of is re-create the list from the object:

The downsides of that particular approach is that 1) it's not self-documenting which list contains what, and more importantly, 2) it scales poorly for additional nested props. The current list of props can be seen [here](https://github.com/mui-org/material-ui/blob/b06c284d830d39fb41f79fb40d3904fd1f67a12f/packages/material-ui-lab/src/Autocomplete/Autocomplete.d.ts#L29-L38). As you can see, there are three nested props, and only one of them is explicitly typed with sub-props.

**Comment by Zaid-Ajaj** [comment]
> The downsides of that particular approach is that 1) it's not self-documenting which list contains what, and more importantly, 2) it scales poorly for additional nested props. The current list of props can be seen here. As you can see, there are three nested props (only one of them is explicitly typed with sub-props).

I see, then I guess another simpler option is to map the `RenderInputParams` one-to-one to an interface? Though I am not sure this helps because the original interface itself is a bit misleading (e.g. `inputProps` vs  `InputProps`)
```fs
module autocomplete = 
  type InputProps =
    abstract className: string
    abstract startAdornment : ReactElement
    abstract endAdornment : ReactElement

  type RenderInputParams = 
    abstract ref: Fable.React.IRefValue<HTMLElement option>
    abstract InputLabelProps : IReactProperty list
    abstract inputProps : IReactProperty list
    abstract InputProps : InputProps 

type autocomplete = 
  static member renderText (renderer: autocomplete.RenderInputParams -> ReactElement) = 
    let wrappedRenderer props = 
      let renderInputParams = {
        new RenderInputParams with
          InputLabelProps = createPropertyList (props?InputLabelProps)
          inputProps = createPropertyList (props?inputProps)
          InputProps = unbox (props?InputProps)
          ref = unbox (props?ref)
      }
      renderer renderInputParams 
      
    unbox<IReactProperty> ("renderText", wrappedRenderer)
```
Using this, you won't be able to spread the root object (i.e `{...props}`) but that is fine if the only property you need to spread is the `ref` which you can `yield` separately. As for `inputProps` and `InputLabelProps`, they are spreadable. 

Usage becomes:
```fs
autocomplete.renderInput (fun renderParams ->
  Mui.textField [
    prop.ref renderParams.ref
    textField.label "Foo"
    textField.inputProps [
      yield! renderParams.inputProps
      // other input props
    ]
  ])
```
I don't how I feel about this solution, it is still a bit confusing for users. Also it becomes an entirely different beast if RenderInputParams has even more undocumented props not declared in that interface but then, you could do *as if* `RenderInputParams` itself is `IReactProperty list` using an extension method:
```fs
type RenderInputParams with
  member this.properties() = unbox<IReactProperty list> (createPropertyList this)
``` 
which you can then spread:
```fs
autocomplete.renderInput (fun renderParams ->
  Mui.textField [
    yield! (renderParams.properties())
    textField.label "Foo"
    textField.inputProps [
      yield! renderParams.inputProps
      // other input props
    ]
  ])
```

**Comment by cmeeren** [comment]
Thanks for the suggestion. Seems we're more or less arrived back at my original suggestion, then.

**Comment by cmeeren** [comment]
I think that for the time being, I'll just do this. Does that seem okay?

```f#
[<Erase>]
type PropsObject =
  /// Gets all properties on this object as Feliz properties you can `yield!`.
  [<Emit("Object.entries($0)")>]
  member inline _.felizProps () : IReactProperty [] = jsNative

[<AbstractClass; Erase>]
type AutocompleteRenderInputParams_InputLabelProps =
  inherit PropsObject

[<AbstractClass; Erase>]
type AutocompleteRenderInputParams_InputProps =
  inherit PropsObject

[<AbstractClass; Erase>]
type AutocompleteRenderInputParams_inputProps =
  inherit PropsObject

[<AbstractClass; Erase>]
type AutocompleteRenderInputParams =
  inherit PropsObject
  abstract InputLabelProps: AutocompleteRenderInputParams_InputLabelProps
  abstract InputProps: AutocompleteRenderInputParams_InputProps
  abstract inputProps: AutocompleteRenderInputParams_inputProps
```

I can't imagine anyone needing individual props (e.g. the `ref` or the `className`). They just `yield!` everything at the needed level, and add with their own props.

**Comment by cmeeren** [comment]
Actually, it's not that hard to surface the other props in a Feliz-friendly manner, so perhaps I'll just do this:

```f#
[<Erase>]
type PropsObject =
  /// Gets all properties on this object as Feliz properties you can `yield!`.
  [<Emit("Object.entries($0)")>]
  member inline _.felizProps : IReactProperty [] = jsNative

[<AbstractClass; Erase>]
type AutocompleteRenderInputParams_InputLabelProps =
  inherit PropsObject

[<AbstractClass; Erase>]
type AutocompleteRenderInputParams_InputProps =
  inherit PropsObject
  [<Emit("[\"className\", $0.className]")>]
  member inline _.className : IReactProperty = jsNative
  [<Emit("[\"startAdornment\", $0.startAdornment]")>]
  member inline _.startAdornment : IReactProperty = jsNative
  [<Emit("[\"endAdornment\", $0.endAdornment]")>]
  member inline _.endAdornment : IReactProperty = jsNative

[<AbstractClass; Erase>]
type AutocompleteRenderInputParams_inputProps =
  inherit PropsObject

[<AbstractClass; Erase>]
type AutocompleteRenderInputParams =
  inherit PropsObject
  abstract InputLabelProps: AutocompleteRenderInputParams_InputLabelProps
  abstract InputProps: AutocompleteRenderInputParams_InputProps
  abstract inputProps: AutocompleteRenderInputParams_inputProps
  [<Emit("[\"ref\", $0.ref]")>]
  member inline _.ref : IReactProperty = jsNative

[<AbstractClass; Erase>]
type AutocompleteRenderGroupParams =
  inherit PropsObject
```

**Comment by Zaid-Ajaj** [comment]
Looks good to me!

**Comment by cmeeren** [comment]
Thanks a lot for the help! 👍 Will be released in 0.8.0 shortly.
